src/H5FDsubfiling/H5FDsubfiling.c

```diff
@@ -1535,199 +1565,16 @@ static herr_t
 H5FD__subfiling_read(H5FD_t *_file, H5FD_mem_t type, hid_t H5_ATTR_UNUSED dxpl_id, haddr_t addr, size_t size,
                      void *buf /*out*/)
 {
-    subfiling_context_t *sf_context         = NULL;
-    H5FD_subfiling_t    *file_ptr           = (H5FD_subfiling_t *)_file;
-    H5FD_mem_t          *io_types           = NULL;
-    haddr_t             *io_addrs           = NULL;
-    size_t              *io_sizes           = NULL;
-    void               **io_bufs            = NULL;
-    int64_t             *source_data_offset = NULL;
-    int64_t             *sf_data_size       = NULL;
-    int64_t             *sf_offset          = NULL;
-    bool                 rank0_bcast        = false;
-    int                  num_subfiles;
-    herr_t               ret_value = SUCCEED;
+    H5FD_subfiling_t *file_ptr  = (H5FD_subfiling_t *)_file;
+    herr_t            ret_value = SUCCEED;
 
-    assert(file_ptr && file_ptr->pub.cls);
+    assert(file_ptr);
+    assert(file_ptr->pub.driver_id == H5FD_SUBFILING);
     assert(buf);
 
-    /* Check for overflow conditions */
-    if (!H5_addr_defined(addr))
```

Comment:
All the logic between read and write was mostly duplicated, so I moved it into `H5FD__subfiling_io_helper`